### ISSUE TITLE
Better HSX error messages

### ISSUE THREAD

**Issue by mpscholten** [digitallyinduced/ihp#916]
As discussed in the last IHP contributors meeting the HSX error messages are terrible.

Ideas:
- Maybe catch the common error messages and show something nicer
- Partially applied function
- Link to an action and not imported the routing file
- Print the HTML surrounding the error context

**Comment by unhammer** [comment]
I just had one, not sure if it's covered by the above:

```
Overlapping instances for IHP.HSX.ToHtml.ToHtml a0
        arising from a use of ‘toHtml’
      Matching instances:
        instance [overlappable] Show a => IHP.HSX.ToHtml.ToHtml a
          -- Defined in ‘IHP.HSX.ToHtml’
        instance IHP.HSX.ToHtml.ToHtml ByteString
          -- Defined in ‘IHP.HSX.ToHtml’
        instance IHP.HSX.ToHtml.ToHtml Text -- Defined in ‘IHP.HSX.ToHtml’
        ...plus two others
        ...plus two instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
      (The choice depends on the instantiation of ‘a0’
       To pick the first instance above, use IncoherentInstances
       when compiling the other instance declarations)
    • In the expression: toHtml sample.userId
      In the fourth argument of ‘IHP.HSX.QQ.makeParent’, namely
        ‘[Text.Blaze.Internal.preEscapedText (Data.Text.pack "Kunde: "),
          toHtml sample.userId]’
      In the first argument of ‘IHP.HSX.QQ.applyAttributes’, namely
        ‘((((IHP.HSX.QQ.makeParent
               (IHP.HSX.QQ.textToStaticString (Data.Text.pack "p")))
              (IHP.HSX.QQ.textToStaticString (Data.Text.pack "<p")))
             (IHP.HSX.QQ.textToStaticString (Data.Text.pack "</p>")))
            [Text.Blaze.Internal.preEscapedText (Data.Text.pack "Kunde: "),
             toHtml sample.userId])’
  |
9 |     html ShowView { .. } = [hsx|
```

where the actual error was doing the equivalent of `<p>post.userId</p>` instead of `<p>post.authorId.userId</p>` (for a type `Include "authorId" Post`)

**Comment by mpscholten** [comment]
Thx for sharing, good example 👍

**Comment by unhammer** [comment]
Is this related to how errors inside hsx blocks always mark the whole block, or is that a separate issue? E.g. if I use an undefined variable deep inside `[hsx|…{myundefined}…|]` then HLS will underline the whole block, instead of just underlining `myundefined`.

**Comment by mpscholten** [comment]
Yes 👍  I think this is the most common case. We should also improve the HSX parser errors (e.g. when a closing tag is missing somewhere right now it's hard to figure out where).

**Comment by kaf-lamed-beyt** [comment]
Hi @mpscholten, I understand that this issue is somewhat related to good error messages. 

Do you mind pointing me to a starting point? I'd love to work on this.

**Comment by Bhavyajain21** [comment]
@mpscholten If it is still open. Can I be assigned?

**Comment by janat08** [comment]
@mpscholten Is learning Haskell over this worthwhile, with more bounties after this? The 100$ is just for the part that OP complained about, and not every error message in HSX?

**Comment by IsxImattI** [comment]
/attempt #916

Plan:
- Hook into the HSX parser to improve error messages for common issues:
  - Undefined variables
  - Type errors in `toHtml` (e.g. overlapping instances)
  - Missing or mismatched tags
- Show surrounding HTML for better context
- Narrow down error location (not the whole [hsx|…|] block)

**Comment by hoklims** [comment]
/attempt #916

Hello, I'd like to work on this issue to improve HSX error messages.

Having reviewed the previous attempts (#2065 by @IsxImattI and #2079 by @luffy-orf), I see they proposed good improvements but lacked demonstration and testing: PR #2065 is still open without visual output or feedback, and #2079 was closed after maintainers requested screenshots of how the errors look. This suggests the maintainers need clear examples of the improved error messages.

My plan is to:
- Identify common HSX errors (e.g. undefined variables, missing `ToHtml` instances, invalid or mismatched tags) and implement hooks in the HSX parser to show a more precise location and the surrounding HTML context, with actionable suggestions.
- Write tests for these scenarios and include a demo (video or screenshots) to show how the new messages look in practice.
- Ask for clarifications as needed and follow the contribution guidelines.

Please let me know if there are additional cases you'd like me to cover. I look forward to improving the developer experience.

**Comment by codygman** [comment]
I see a bounty is still open for this on Aglora, are you still looking for a solution @mpscholten ? Is the linked open PR not viable for some reason.

**Comment by jacanidjar** [comment]
@mpscholten This bounty is available, can I start it?
